### DISCUSSION TITLE
Speaker selection function (state_transition) with "parallel" agents processing

### DISCUSSION THREAD

**Discussion by amarquinar** [microsoft/autogen#4415] [Q&A]
Usually the state transition returns just a one agent. For example, the following code means that the workflow will be executed in sequence, first initializer, then coder, then executor, then scientist and finally data_analyst. In that case the output data of executor is modified by scientist agent and data_analyst agent receive the data with different format. 

```
...
def state_transition(last_speaker, groupchat):
    messages = groupchat.messages

    if last_speaker is initializer:
        return coder
    elif last_speaker is coder:
        return executor
    elif last_speaker is executor:
        return scientist
    elif last_speaker is scientist:
        return data_analyst
    elif last_speaker is data_analyst:
        return None
....
```
We need that the data_analyst receive the output from the executor with the original format. 

Is it possible that, for example, the output of executor will being send to scientist and data_analyst an the same time? We need that the output of executer will be read by the both agents (scientist and data_analyst) with the same format.

How do we need to define the state_transition function for this case?

Many thanks in advance!

**Comment by ekzhu** [comment]
If you don't care about strict parallelism, the agents inside a group chat already share the same message context. So, the output of the executor is already available for every agent in the group chat. So you could perhaps prompt the scientist and analyst agents to only use the original output of the code executor. 

If you are looking for a lot more control over exactly how agent behaves and it looks like you have a well-defined workflow, I recommend you take a look at the v0.4 Core API and implement a pub-sub based workflow that enables true parallelism.

**Reply by amarquinar** [reply]
First of all, many thanks for replying quickly @ekzhu !

Could you please share a quick example about how to implement a pub-sub workflow that enables true parallelism?

Thanks again!

**Reply by ekzhu** [reply]
We can provide a simple example (https://github.com/microsoft/autogen/issues/4427). However, in the meantime I recommend reading the Core API documentation and learn about the basics.

**Reply by amarquinar** [reply]
Many thanks for your time @ekzhu
